modules/data-streams/src/main/java/org/elasticsearch/datastreams/DataStreamIndexSettingsProvider.java

```diff
@@ -161,19 +164,34 @@ private List<String> findRoutingPaths(String indexName, Settings allSettings, Li
 
             List<String> routingPaths = null;
             for (var fieldMapper : mapperService.documentMapper().mappers().fieldMappers()) {
-                if (fieldMapper instanceof KeywordFieldMapper keywordFieldMapper) {
-                    if (keywordFieldMapper.fieldType().isDimension()) {
-                        if (routingPaths == null) {
-                            routingPaths = new ArrayList<>();
-                        }
-                        routingPaths.add(keywordFieldMapper.name());
-                    }
-                }
+                routingPaths = extractPath(routingPaths, fieldMapper);
+            }
+            for (var template : mapperService.getAllDynamicTemplates()) {
+                var templateName = "__dynamic__" + template.name();
+                var mappingSnippet = template.mappingForName(templateName, KeywordFieldMapper.CONTENT_TYPE);
+
+                MappingParserContext parserContext = mapperService.parserContext();
+                var mapper = parserContext.typeParser(KeywordFieldMapper.CONTENT_TYPE)
+                    .parse(template.pathMatch(), mappingSnippet, parserContext)
+                    .build(MapperBuilderContext.ROOT);
+                routingPaths = extractPath(routingPaths, mapper);
             }
             return routingPaths;
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
+    private static List<String> extractPath(List<String> routingPaths, Mapper mapper) {
```

Comment:
I really don't like that this method takes both a `routingPaths` parameter (which it mutates) and returns the `routingPaths` object. Can we make it either take `routingPaths` and mutate it but then return `void` (and add javadocs), or else make it not mutate the argument and return a new list?

Reply:
Prior to this change the `routingPaths` list was lazily initialised in the `findRoutingPaths(...)` method.
Similar logic was used for extracting paths from dynamic templates, so I generated this method.

Maybe it just easier to remove the lazy initialising of the list part and then the function can return void and update the provided list, which is your first suggestion.